### PR TITLE
fix: #36 - retain playback rate in all situations

### PR DESCRIPTION
**Pull request by fivecar**
The call to `avPlayer.play()` resets the playback rate to 1.0 (according to official docs, and referred to [here](https://stackoverflow.com/questions/8688872/)). People have been saying in `react-native-track-player` for years that this has been happening (e.g. https://github.com/doublesymmetry/react-native-track-player/issues/614).

This fixes that. We now store an internal expected playback rate (`playRate`), and instead of calling `avPlayer.play()`, we simply restore the rate to what the caller expects.

Tested this by using a bare-minimum [repro project](https://github.com/SuperphonicHub/TrackPlayerRateExample) and modifying the Swift package directly to verify that the bug is fixed.

Sorry about the whitespace removal... I think that's my VS Code talking.

### CODE REVIEW

**Comment by fivecar** [comment]
@dcvz I suspect the workflow is blocked on your approval... would you mind unblocking it, and letting me know what you think of this PR? Thank you!

**Comment by jspizziri** [comment]
@fivecar please update your PR to remove styling changes and only contain the fix for the issue at hand. That would make review a bit easier.

**Comment by fivecar** [comment]
Sure, @jspizziri . I had originally thought it would be simple to review by using:

image

But now I've rolled back and redone the change as you've requested.

File: SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift [hunk 1]
```diff
@@ -156,7 +162,12 @@ class AVPlayerWrapper: AVPlayerWrapperProtocol {
     
     func play() {
         playWhenReady = true
-        avPlayer.play()
+
+        // You might be tempted to avPlayer.play() here, but that essentially
```
Comment (jspizziri) [inline SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift:166]: I don't have a deep knowledge of AVPlayer, but removing `avPlayer.play()` here seems like it would fail to play an audio that is in the paused state, no?

File: SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift [hunk 2]
```diff
@@ -37,6 +37,9 @@ class AVPlayerWrapper: AVPlayerWrapperProtocol {
 
     /// True when the track was paused for the purpose of switching tracks
     fileprivate var pausedForLoad: Bool = false
+    // We need to track this ourselves, in addition to avPlayer having its rate,
+    // because any call to avPlayer.play() resets its playback rate to 0!
```
Comment (jspizziri) [inline SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift:41]: The linked stackoverflow says it resets the rate to 1.0, not 0. Also, is there official Apple documentation about this? It seems very strange that they'd not track the rate internally. I think having a test to go along with this is critical.

**Comment by fivecar** [comment]
> The only other thing I think the team will want to see is an accompanying test that demonstrates the issue/resolution. We've been trying to be more diligent about adding tests recently.
@jspizziri : it appears (at least to me, running off commit 23fdb9b9db, "Release 0.15.3"), that the tests already fail. Specifically, test_AudioPlayer__state__stopping_source__should_be_idle fails, at least on my machine. I'm guessing the tests are already running via automation on the main branch, so I'm not sure what I'm doing wrong.

File: SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift (hunk 2 above)
Comment (fivecar) [reply SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift:41]: My amateur-hour understanding of AVWrapper is really bad, but it seems at least somewhere that it expects the rate [to be made zero](https://github.com/doublesymmetry/SwiftAudioEx/blob/23fdb9b9db4b759cf99390ae8c5381dc7a21a0ae/Example/Tests/AVPlayerWrapperTests.swift#L171).

@jspizziri : I'm no expert, but take a look at [this code from AVPlayer](https://github.com/xybp888/iOS-SDKs/blob/77a0d14fde495bb1e74a857527b77de4090230db/iPhoneOS10.3.sdk/System/Library/Frameworks/AVFoundation.framework/Headers/AVPlayer.h#L139). I believe the years-old bug in SwiftAudioEx might be due to one of these edge cases being hit, where AVPlayer resets the rate.

The tests otherwise all pass (meaning setting the rate truly does play the audio).

File: SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift (hunk 1 above)
Comment (fivecar) [reply SwiftAudioEx/Classes/AVPlayerWrapper/AVPlayerWrapper.swift:166]: (see comment below; and also, tests otherwise pass [i.e. playback definitely resumes, according to the tests; and also according to integrating this change into my actual app]).

**Comment by jspizziri** [comment]
@fivecar, congrats on your first Swift code 🎊 . Also thanks for all the work and investigation here.

Have you looked at any of the work that @puckey [has done here](https://github.com/doublesymmetry/SwiftAudioEx/pull/28)? The goal is for that to get merged soon and there's been a lot of work done. Some of it looks like it might've touched what you've worked on.

I'd recommend looking at the code, and if you want to take it for a spin I can probably help you with that (depends on how you're installing it in your project). All in all I think I'd like @puckey to weigh in on the above PR as I'm only slightly familiar with AVPlayer.

**Comment by fivecar** [comment]
@jspizziri : Ah, indeed, @puckey's change there addresses the same concern (specifically, that `_rate` should be memorized and reapplied every time you `play()`).

That changes has been 3+ months in the works; hoping that'll land soon.

Going to abandon this one in the meantime. I've got it patched on my own machine, so I'm unblocked right now.
